src/hotspot/share/nmt/virtualMemoryTracker.cpp

```diff
@@ -47,545 +47,211 @@ void VirtualMemory::update_peak(size_t size) {
 
 void VirtualMemorySummary::snapshot(VirtualMemorySnapshot* s) {
   // Snapshot current thread stacks
-  VirtualMemoryTracker::snapshot_thread_stacks();
+  VirtualMemoryTracker::Instance::snapshot_thread_stacks();
   as_snapshot()->copy_to(s);
 }
 
-SortedLinkedList<ReservedMemoryRegion, compare_reserved_region_base>* VirtualMemoryTracker::_reserved_regions;
-
-int compare_committed_region(const CommittedMemoryRegion& r1, const CommittedMemoryRegion& r2) {
-  return r1.compare(r2);
-}
-
-int compare_reserved_region_base(const ReservedMemoryRegion& r1, const ReservedMemoryRegion& r2) {
-  return r1.compare(r2);
-}
-
-static bool is_mergeable_with(CommittedMemoryRegion* rgn, address addr, size_t size, const NativeCallStack& stack) {
-  return rgn->adjacent_to(addr, size) && rgn->call_stack()->equals(stack);
-}
-
-static bool is_same_as(CommittedMemoryRegion* rgn, address addr, size_t size, const NativeCallStack& stack) {
-  // It would have made sense to use rgn->equals(...), but equals returns true for overlapping regions.
-  return rgn->same_region(addr, size) && rgn->call_stack()->equals(stack);
-}
-
-static LinkedListNode<CommittedMemoryRegion>* find_preceding_node_from(LinkedListNode<CommittedMemoryRegion>* from, address addr) {
-  LinkedListNode<CommittedMemoryRegion>* preceding = nullptr;
-
-  for (LinkedListNode<CommittedMemoryRegion>* node = from; node != nullptr; node = node->next()) {
-    CommittedMemoryRegion* rgn = node->data();
-
-    // We searched past the region start.
-    if (rgn->end() > addr) {
-      break;
-    }
-
-    preceding = node;
+bool VirtualMemoryTracker::Instance::initialize(NMT_TrackingLevel level) {
+  assert(_tracker == nullptr, "only call once");
+  if (level >= NMT_summary) {
+    void* tracker = os::malloc(sizeof(VirtualMemoryTracker), mtNMT);
+    if (tracker == nullptr) return false;
+    _tracker = new (tracker) VirtualMemoryTracker(level == NMT_detail);
+    return _tracker->tree() != nullptr;
```

Comment:
@afshin-zafari , `_tracker->tree()` can never be null anymore. In the future we should do a PR where we change it to return a reference.

Reply:
Removed.